File: app/src/main/java/com/android/example/popularmovies/ui/detail/fragments/reviews/ReviewsFragment.java
package com.android.example.popularmovies.ui.detail.fragments.reviews;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.android.example.popularmovies.LoadingStatus;
import com.android.example.popularmovies.R;
import com.android.example.popularmovies.database.FavMovieEntry;
import com.android.example.popularmovies.databinding.FragmentReviewsBinding;
import com.android.example.popularmovies.ui.detail.fragments.BaseDetailFragment;
import com.android.example.popularmovies.utils.json.MovieReview;

public class ReviewsFragment extends BaseDetailFragment implements ReviewsAdapter.ReviewOnClickListener {

    private ReviewsViewModel viewModel;
    private FragmentReviewsBinding binding;
    private ReviewsAdapter adapter;

    public static ReviewsFragment newInstance() {
        return new ReviewsFragment();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = FragmentReviewsBinding.inflate(inflater, container, false);

        LinearLayoutManager layoutManager = new LinearLayoutManager(requireContext());
        binding.rvReviewList.setLayoutManager(layoutManager);

        adapter = new ReviewsAdapter(this);
        binding.rvReviewList.setAdapter(adapter);

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(this).get(ReviewsViewModel.class);
        viewModel.getStatus().observe(getViewLifecycleOwner(), status -> {
            switch (status){
                case ERROR:
                    binding.tvErrorText.setVisibility(View.VISIBLE);
                    binding.llReviewLayout.setVisibility(View.INVISIBLE);
                    binding.progressBar.setVisibility(View.INVISIBLE);
                    break;
                case LOADING:
                    binding.progressBar.setVisibility(View.VISIBLE);
                    binding.tvErrorText.setVisibility(View.INVISIBLE);
                    binding.llReviewLayout.setVisibility(View.INVISIBLE);
                    break;
                case DONE:
                    binding.llReviewLayout.setVisibility(View.VISIBLE);
                    binding.tvErrorText.setVisibility(View.INVISIBLE);
                    binding.progressBar.setVisibility(View.INVISIBLE);
                    break;
            }
        });
        viewModel.getReviews().observe(getViewLifecycleOwner(), movieReviews -> {
            if (movieReviews != null && !movieReviews.isEmpty()) {
                adapter.setReviews(movieReviews);
            }
        });
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        FavMovieEntry movie = getMovie();
        if (movie != null) {
            viewModel.fetchReviews(getString(R.string.api_key_v3), movie.getId());
        } else {
            viewModel.setStatus(LoadingStatus.ERROR);
            Log.e(getClass().getSimpleName(), "Null movie passed to fragment.");
        }
    }

    @Override
    public void onListItemClick(MovieReview review) {
        Intent intent = new Intent(Intent.ACTION_VIEW, review.getUrlAsUri())
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        if (intent.resolveActivity(requireActivity().getPackageManager()) != null) {
            startActivity(intent);
        }
    }
}

File: app/src/main/java/com/android/example/popularmovies/ui/detail/fragments/reviews/ReviewsFragment.kt
package com.android.example.popularmovies.ui.detail.fragments.reviews

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.android.example.popularmovies.R
import com.android.example.popularmovies.databinding.FragmentReviewsBinding
import com.android.example.popularmovies.ui.detail.fragments.BaseDetailFragment
import com.android.example.popularmovies.ui.detail.fragments.reviews.ReviewsAdapter.ReviewOnClickListener
import com.android.example.popularmovies.utils.json.MovieReview
import com.android.example.popularmovies.utils.network.LoadingStatus

class ReviewsFragment : BaseDetailFragment(), ReviewOnClickListener {
    private val viewModel by viewModels<ReviewsViewModel>()
    private lateinit var binding: FragmentReviewsBinding
    private lateinit var adapter: ReviewsAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentReviewsBinding.inflate(inflater, container, false)
        val layoutManager = LinearLayoutManager(requireContext())
        binding.rvReviewList.layoutManager = layoutManager
        adapter = ReviewsAdapter(requireContext(), this)
        binding.rvReviewList.adapter = adapter
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel.status.observe(viewLifecycleOwner, Observer { status: LoadingStatus? ->
            when (status) {
                LoadingStatus.ERROR -> {
                    binding.tvErrorText.visibility = View.VISIBLE
                    binding.llReviewLayout.visibility = View.INVISIBLE
                    binding.progressBar.visibility = View.INVISIBLE
                }
                LoadingStatus.LOADING -> {
                    binding.progressBar.visibility = View.VISIBLE
                    binding.tvErrorText.visibility = View.INVISIBLE
                    binding.llReviewLayout.visibility = View.INVISIBLE
                }
                LoadingStatus.DONE -> {
                    binding.llReviewLayout.visibility = View.VISIBLE
                    binding.tvErrorText.visibility = View.INVISIBLE
                    binding.progressBar.visibility = View.INVISIBLE
                }
            }
        })
        viewModel.reviews.observe(viewLifecycleOwner, Observer { movieReviews: List<MovieReview>? ->
            if (movieReviews != null && movieReviews.isNotEmpty()) {
                adapter.setReviews(movieReviews)
            }
        })
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val movie = movie
        if (movie != null) {
            viewModel.fetchReviews(getString(R.string.api_key_v3), movie.id)
        } else {
            viewModel.setStatus(LoadingStatus.ERROR)
            Log.e(javaClass.simpleName, "Null movie passed to fragment.")
        }
    }

    override fun onListItemClick(review: MovieReview?) {
        val intent = Intent(Intent.ACTION_VIEW, review?.urlAsUri)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        if (intent.resolveActivity(requireActivity().packageManager) != null) {
            startActivity(intent)
        }
    }

    companion object {
        fun newInstance(): ReviewsFragment {
            return ReviewsFragment()
        }
    }
}